pkg/runtime/adoption_reconciler_test.go

```diff
@@ -0,0 +1,505 @@
+
+	rmo.On("GetLabels").Return(make(map[string]string))
+	rmo.On("GetAnnotations").Return(make(map[string]string))
+	rmo.On("GetFinalizers").Return(make([]string, 0))
+	rmo.On("GetOwnerReferences").Return(make([]v1.OwnerReference, 0))
+	rmo.On("GetGenerateName").Return("")
+}
+
+func setupMockManager(manager *ackmocks.AWSResourceManager, ctx context.Context, res *ackmocks.AWSResource) {
+	manager.On("ReadOne", ctx, res).Return(res, nil)
+}
+
+func setupMockDescriptor(descriptor *ackmocks.AWSResourceDescriptor, res *ackmocks.AWSResource) {
+	descriptor.On("MarkManaged", res).Run(func(args mock.Arguments) {})
+	descriptor.On("MarkAdopted", res).Run(func(args mock.Arguments) {})
+}
+
+func setupMockApiReader(apiReader *ctrlrtclientmock.Reader, ctx context.Context, res *ackmocks.AWSResource) {
+	apiReader.On("Get", ctx, types.NamespacedName{
+		Namespace: Namespace,
+		Name:      Name,
+	},res.RuntimeObject()).Return(k8serrors.NewNotFound(schema.GroupResource{}, ""))
+}
+
+func adoptedResource(namespace , name string) *ackv1alpha1.AdoptedResource {
+	return &ackv1alpha1.AdoptedResource{
+		TypeMeta:   v1.TypeMeta{},
+		ObjectMeta: v1.ObjectMeta{
+			Namespace:namespace,
+			Name:name,
+		},
+		Spec:       ackv1alpha1.AdoptedResourceSpec{
+			Kubernetes: nil,
+			AWS:        &ackv1alpha1.AWSIdentifiers{NameOrID: "name"},
+		},
+		Status:     ackv1alpha1.AdoptedResourceStatus{},
+	}
+}
+
+//Tests
+
+func TestSync_FailureInSettingIdentifiers(t *testing.T) {
+	// Setup
+	require := require.New(t)
+	// Mock resource creation
+	r, kc, apiReader := mockReconciler()
+	descriptor, res := mockDescriptorAndAWSResource()
+	manager := mockManager()
+	adoptedRes := adoptedResource(Namespace, Name)
+	res.On("SetIdentifiers", adoptedRes.Spec.AWS).Return(errors.New("unable to set Identifier"))
+	ctx := context.TODO()
+	statusWriter := &ctrlrtclientmock.StatusWriter{}
+
+	//Mock behavior setup
+	setupMockAwsResource(res, adoptedRes)
+	setupMockClient(kc, statusWriter, ctx, adoptedRes)
+
+	// Call
+	err := r.Sync(ctx, descriptor, manager, adoptedRes)
+
+	// Assertions
+	// error occured
+	require.NotNil(err)
+	require.Equal("unable to set Identifier", err.Error())
+	// Identifiers are set from AdoptedResource into AWSResource
+	res.AssertCalled(t, "SetIdentifiers", adoptedRes.Spec.AWS)
+	// ReadOne call is not made to find observed state of AWSResource
+	manager.AssertNotCalled(t, "ReadOne", ctx, res)
+	// No calls to findout if the AWSResource already exists
+	apiReader.AssertNotCalled(t, "Get", ctx, types.NamespacedName{
+		Namespace: Namespace,
+		Name:      Name,
+	}, res.RuntimeObject())
+	// Create AWSResource should not happen
+	kc.AssertNotCalled(t, "Create", ctx, res.RuntimeObject())
+	// Update status of AWSResource should not happen
+	statusWriter.AssertNotCalled(t, "Update", ctx, res.RuntimeObject())
+	// AdoptedResource did not get marked as managed
+	kc.AssertNotCalled(t, "Patch", ctx, adoptedRes, mock.AnythingOfType("*client.mergeFromPatch"))
```

Comment:
Hmm as per your code comment above. This should probably be marked as managed by this point, because we have already attempt to manage it - even though it was unsuccessful.

Reply:
Yeah, that's a bug in the actual code. I have added a TODO for it in the source.
In this PR, i will just add test for whatever functionality is there as is.